In [2]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

In [4]:
mnist_base_path="/Users/zhouwencheng/Desktop/Grass/data/picture/mnist"
# 载入数据集
mnist = input_data.read_data_sets(mnist_base_path, one_hot=True)

Extracting ./data/MNIST_data/train-images-idx3-ubyte.gz
Extracting ./data/MNIST_data/train-labels-idx1-ubyte.gz
Extracting ./data/MNIST_data/t10k-images-idx3-ubyte.gz
Extracting ./data/MNIST_data/t10k-labels-idx1-ubyte.gz


In [17]:
# 每个批次的大小
batch_size = 100
# 计算一共有多少个批次 (550)
n_batch = mnist.train.num_examples // batch_size

# 定义两个 placeholder
x = tf.placeholder(tf.float32, [None, 784])
y = tf.placeholder(tf.float32, [None, 10])
keep_prob = tf.placeholder(tf.float32)

# 创建一个简单的神经网络
W1 = tf.Variable(tf.truncated_normal([784, 200], stddev = 0.1))
b1 = tf.Variable(tf.zeros([200]) + 0.1)
L1 = tf.nn.tanh(tf.matmul(x, W1) + b1)
L1_drop = tf.nn.dropout(L1, keep_prob)

W2 = tf.Variable(tf.truncated_normal([200, 200], stddev = 0.1))
b2 = tf.Variable(tf.zeros([200]) + 0.1)
L2 = tf.nn.tanh(tf.matmul(L1_drop, W2) + b2)
L2_drop = tf.nn.dropout(L2, keep_prob)

W3 = tf.Variable(tf.truncated_normal([200, 100], stddev = 0.1))
b3 = tf.Variable(tf.zeros([100]) + 0.1)
L3 = tf.nn.tanh(tf.matmul(L2_drop, W3) + b3)
L3_drop = tf.nn.dropout(L3, keep_prob)

W4 = tf.Variable(tf.truncated_normal([100, 10], stddev = 0.1))
b4 = tf.Variable(tf.zeros([10]) + 0.1)
prediction = tf.nn.softmax(tf.matmul(L3_drop, W4) + b4)

# 定义 二次代价函数
# loss = tf.reduce_mean(tf.square(y - prediction))
# 使用交叉熵
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels = y, logits = prediction))

# 使用梯度下降法
train_step = tf.train.GradientDescentOptimizer(0.2).minimize(loss)
# 初始化变量
init = tf.global_variables_initializer()

# 结果放在一个布尔型列表中
correct_prediction = tf.equal(tf.arg_max(y, 1), tf.arg_max(prediction, 1))
# 求准确率
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
    sess.run(init)
    for epoch in range(21):
        for batch in range(n_batch):
            batch_xs, batch_ys = mnist.train.next_batch(batch_size)
            sess.run(train_step, feed_dict = {x:batch_xs, y:batch_ys, keep_prob:0.8})
        acc = sess.run(accuracy, feed_dict = {x:mnist.test.images, y:mnist.test.labels, keep_prob:1.0})
        acc1 = sess.run(accuracy, feed_dict = {x:mnist.train.images, y:mnist.train.labels, keep_prob:1.0})
        print('accuracy:', acc, ' train acc:', acc1)
            


n_batch

accuracy: 0.9091  train acc: 0.90154546
accuracy: 0.9221  train acc: 0.9200909
accuracy: 0.9307  train acc: 0.92938185
accuracy: 0.9361  train acc: 0.93683636
accuracy: 0.9398  train acc: 0.9411455
accuracy: 0.9434  train acc: 0.94574547
accuracy: 0.9446  train acc: 0.9491818
accuracy: 0.947  train acc: 0.9518545
accuracy: 0.9502  train acc: 0.9538182
accuracy: 0.951  train acc: 0.95594543
accuracy: 0.9525  train acc: 0.9578909
accuracy: 0.9564  train acc: 0.9604727
accuracy: 0.9562  train acc: 0.9620182
accuracy: 0.9573  train acc: 0.9632909
accuracy: 0.959  train acc: 0.9650364
accuracy: 0.9589  train acc: 0.9653091
accuracy: 0.9601  train acc: 0.9670727
accuracy: 0.9615  train acc: 0.96834546
accuracy: 0.9621  train acc: 0.96903634
accuracy: 0.9606  train acc: 0.96863633
accuracy: 0.9642  train acc: 0.97163635


550



    drop = 1.0

    accuracy: 0.9186  train acc: 0.91456366
    accuracy: 0.9351  train acc: 0.9356545
    accuracy: 0.9434  train acc: 0.9452364
    accuracy: 0.95  train acc: 0.9541091
    accuracy: 0.9538  train acc: 0.9592182
    accuracy: 0.9566  train acc: 0.96352726
    accuracy: 0.9578  train acc: 0.96738183
    accuracy: 0.9625  train acc: 0.96901816
    accuracy: 0.9637  train acc: 0.9735818
    accuracy: 0.9653  train acc: 0.9751818
    accuracy: 0.9664  train acc: 0.9776
    accuracy: 0.9681  train acc: 0.97967273
    accuracy: 0.9679  train acc: 0.9803454
    accuracy: 0.9691  train acc: 0.982
    accuracy: 0.9678  train acc: 0.98267275
    accuracy: 0.9694  train acc: 0.98434544
    accuracy: 0.9705  train acc: 0.9843818
    accuracy: 0.9713  train acc: 0.98585457
    accuracy: 0.9711  train acc: 0.985
    accuracy: 0.9721  train acc: 0.98701817
    accuracy: 0.9729  train acc: 0.98767275
